uxarray/helpers.py

```diff
@@ -89,3 +91,275 @@ def determine_file_type(filepath):
+
+    if lon_pt < 0.0:
+        raise Exception('lon_pt out of range ( {} < 0)"'.format(lon_pt))
+
+    if lon_pt > old_lon_width:
+        raise Exception('lon_pt out of range ( {} > {})"'.format(lon_pt, old_lon_width))
+
+    # Expand latitudes
+    if lat_pt > latlon_box[0][1]:
+        latlon_box[0][1] = lat_pt
+
+    if lat_pt < latlon_box[0][0]:
+        latlon_box[0][0] = lat_pt
+
+    # Expand longitude, if non-periodic
+    if not is_lon_periodic:
+        if lon_pt > latlon_box[1][1]:
+            latlon_box[1][1] = lon_pt
+        if lon_pt < latlon_box[1][0]:
+            latlon_box[1][0] = lon_pt
+        return
+
+    # New longitude lies within existing range
+    if latlon_box[1][0] <= latlon_box[1][1]:
+        if lon_pt >= latlon_box[1][0] and lon_pt <= latlon_box[1][1]:
+            return
+    else:
+        if lon_pt >= latlon_box[1][0] or lon_pt <= latlon_box[1][0]:
+            return
+
+    # New longitude lies outside of existing range
+    box_a = latlon_box
+    box_a[1][0] = lon_pt
+
+    box_b = latlon_box
+    box_b[1][1] = lon_pt
+
+    # The updated box is the box of minimum width
+    d_width_now = get_latlonbox_width(latlon_box)
+    d_width_a = get_latlonbox_width(box_a)
+    d_width_b = get_latlonbox_width(box_b)
+
+    if (d_width_a - d_width_now) < -1.0e-14 or (d_width_b - d_width_now) < -1.0e-14:
+        raise Exception('logic error')
+
+    if d_width_a < d_width_b:
+        return box_a
+    else:
+        return box_b
+
+
+
+# helper function to calculate the latlonbox width
+def get_latlonbox_width(latlonbox, is_lon_periodic):
+    """Calculate the width of this LatLonBox
+    Parameters: latlonbox: float array, lat lon box [[lat_0, lat_1],[lon_0, lon_1]],required
+                is_lon_periodic: boolean, Flag indicating the latlonbox is a regional (default to be True).
+
+    Returns: float array, a lat lon box [[lat_0, lat_1],[lon_0, lon_1]]
+
+    Raises:
+       Exception: Logic Errors
+    """
+
+    if not is_lon_periodic:
+        return latlonbox[1][1] - latlonbox[1][0]
+
+    if latlonbox[1][0] == latlonbox[1][1]:
+        return 0.0
+    elif latlonbox[1][0] <= latlonbox[1][1]:
+        return latlonbox[1][1] - latlonbox[1][0]
+    else:
+        latlonbox[1][1] - latlonbox[1][0] + (2 * np.pi)
+
+# helper function to calculate latitude and longitude from a node's normalized 3D Cartesian
+# coordinates, in radians.
+def convert_node_XYZ_2_latlon_rad(node_coord):
+    """Calculate the latitude and longitude in radiance for a node represented in the [x, y, z] 3D Cartesian coordinates.
+    Parameters: node_coord: float array, [x, y, z],required
+
+    Returns: float array, [latitude_rad, longitude_rad]
+
+    Raises:
+       Exception: Logic Errors
+    """
+    reference_tolerance = 1.0e-12
+    dx = node_coord[0]
+    dy = node_coord[1]
+    dz = node_coord[2]
+
+    d_mag_2 = dx * dx + dy * dy + dz * dz
+
+    if np.absolute(d_mag_2 - 1.0) >= 0.01:
```

Comment:
sqrt of d_mag_2 needed?

why raise an exception?  rather: just convert to unit magnitude with a warning.

Reply:
Thank you for your information. I will modify it accordingly in the following commits.